src/table.mjs

```diff
@@ -0,0 +1,233 @@
+export const __query = Object.assign(
+  // This function is used by table cells.
+  async (source, operations, invalidation) => {
+    const args = makeQueryTemplate(operations, await source);
+    if (!args) return null; // the empty state
+    return evaluateQuery(await source, args, invalidation);
+  },
+  {
+    // This function is used by SQL cells.
+    sql(source, invalidation) {
+      return async function () {
+        return evaluateQuery(source, arguments, invalidation);
+      };
+    }
+  }
+);
+
+async function evaluateQuery(source, args, invalidation) {
+  if (!source) return;
+
+  // If this DatabaseClient supports abort and streaming, use that.
+  if (typeof source.queryTag === "function") {
+    const abortController = new AbortController();
+    const options = {signal: abortController.signal};
+    invalidation.then(() => abortController.abort("invalidated"));
+    if (typeof source.queryStream === "function") {
+      return accumulateQuery(
+        source.queryStream(...source.queryTag.apply(source, args), options)
+      );
+    }
+    if (typeof source.query === "function") {
+      return source.query(...source.queryTag.apply(source, args), options);
+    }
+  }
+
+  // Otherwise, fallback to the basic sql tagged template literal.
+  if (typeof source.sql === "function") {
+    return source.sql.apply(source, args);
+  }
+
+  // TODO: test if source is a file attachment, and support CSV etc.
+  throw new Error("source does not implement query, queryStream, or sql");
+}
+
+// Generator function that yields accumulated query results client.queryStream
+async function* accumulateQuery(queryRequest) {
+  const queryResponse = await queryRequest;
+  const values = [];
+  values.done = false;
+  values.error = null;
+  values.schema = queryResponse.schema;
+  try {
+    const iterator = queryResponse.readRows();
+    do {
+      const result = await iterator.next();
+      if (result.done) {
+        values.done = true;
+      } else {
+        for (const value of result.value) {
+          values.push(value);
+        }
+      }
+      yield values;
+    } while (!values.done);
+  } catch (error) {
+    values.error = error;
+    yield values;
+  }
+}
+
+/**
+ * Returns a SQL query in the form [[parts], ...params] where parts is an array
+ * of sub-strings and params are the parameter values to be inserted between each
+ * sub-string.
+ */
+ export function makeQueryTemplate(operations, source) {
```

Comment:
we still need this function in Next because we use it to construct the query for the convert to SQL operation. should we also export it as part of the standard library? that feels wrong somehow ... maybe there's another way to share the code?

Reply:
Yes we can have it as an export in src/index.mjs